### PR TITLE
Fixed the 'scale to and from 0' issue

### PR DESCRIPTION
**Pull request by ajanth97**
This [issue](https://github.com/kedacore/http-add-on/issues/214) has now been fixed.

We will now work on making this process more efficient and reducing latency when scaling from 0.

### CODE REVIEW

File: scaler/handlers.go
Comment (khaosdoctor) [inline scaler/handlers.go:58]: If this happens I think the scaler will fail and crash. Can we just set the `hostCount` to 0?

File: scaler/handlers.go
Comment (khaosdoctor) [inline scaler/handlers.go:51]: To avoid this repeated return, what about creating a separated `func` to execute the logic? Like:

```go
func isActiveCheck (allCounts map[string]int, host string) bool {
	if host == "interceptor" {
      return true
    }
    hostCount, ok := allCounts[host]

    return ok && hostCount > 0 ? true : false
}
```

File: scaler/handlers.go
Comment (khaosdoctor) [inline scaler/handlers.go:62]: The you can replace it by:

```go
return & externalscaler.IsActiveResponse{
  Result: isActiveCheck(allCounts, host)
}, nil
```

Comment (khaosdoctor) [reply scaler/handlers.go:62]: And delete most of the logic

File: scaler/handlers.go
Comment (arschles) [inline scaler/handlers.go]: @ajanth97 can you remove this change? It doesn't appear necessary

File: scaler/handlers.go
Comment (arschles) [inline scaler/handlers.go:67]: @ajanth97 can you remove this change? It doesn't appear necessary

File: scaler/handlers.go
Comment (arschles) [reply scaler/handlers.go:58]: The scaler wouldn't crash here, it would just return an error to KEDA. It would be better to keep this logic because if we return an error, KEDA won't take a scaling action. If we return 0, we'll end up telling KEDA to scale the app to zero, which may not be what we want in a failure domain where the scaler doesn't have information about the given host.

File: scaler/handlers.go
Comment (arschles) [inline scaler/handlers.go]:
```suggestion
	active := hostCount > 0
```

File: scaler/handlers.go
Comment (arschles) [inline scaler/handlers.go]: can you log something more descriptive here?

```suggestion
		lggr.Error(err, "given host was not found", "host", host, "allCounts", allCounts)
```

File: scaler/handlers.go
Comment (arschles) [inline scaler/handlers.go]: the second parameter to `lggr.Error` is a message for context, and the ones after that should be in pairs of two, where the first of each pair is the name of the value, and the second is the value. So, the `lggr.Error` call will need to change. Here's something that makes a bit more sense to me. Of course, feel free to change as you see fit.


```suggestion
              err := fmt.Errorf("no 'host' field found in ScaledObject metadata")
		lggr.Error(err, "returning immediately from IsActive RPC call", "ScaledObject", *scaledObject)
		return nil, err
```

Also note that we're dereferencing the `scaledObject` var (prefixing with `*`) because it's passed as a pointer. IIRC, this isn't strictly necessary for the particular implementation of `logr.Logger` that we're using, but it's a best practice that might come in handy if/when we decide to change implementations.

**Review by arschles — CHANGES_REQUESTED** [review]
@ajanth97 this is on a good track. A few more things to do and then we'll be ready to merge.

File: scaler/handlers.go
Comment (ajanth97) [reply scaler/handlers.go:51]: I think it would be a bit over-engineering at the moment. We can do it in the next iteration when we plan to do the push from the interceptor.

File: scaler/handlers.go
Comment (ajanth97) [reply scaler/handlers.go:67]: I added this to maintain consistency between the `StreamIsActive` and `IsActive` methods ? Don't you think that makes more sense to have the same variable name ?

File: scaler/handlers.go
Comment (ajanth97) [reply scaler/handlers.go:51]: Also we wouldn't be calling this function in a lot of places right now right ? So we don't have to abstract this logic right ?

File: scaler/handlers.go
Comment (ajanth97) [reply scaler/handlers.go]: Maybe to be more descriptive, how about :
`lggr.Error(err, "Given host was not found in queue count map", "host", host, "allCounts", allCounts)`

File: scaler/handlers.go
Comment (arschles) [reply scaler/handlers.go:51]: having such an abstraction would only be applicable if we're going to set `hostCount` to zero as a fallback, as detailed in https://github.com/arschles/http-add-on/pull/12/files#r688098164. I think doing so could be harmful in some failure scenarios (elaborated below), but some sort of abstraction which doesn't include setting `hostCount` to zero could in theory be helpful. @khaosdoctor can think of something like this?

File: scaler/handlers.go
Comment (arschles) [reply scaler/handlers.go:67]: the value isn't used in this function so it doesn't really matter what this variable name is called, so that's why I erred on the side of making fewer changes. I'll leave it up to you here.

**Review by arschles — APPROVED** [review]
@ajanth97 I see that there are still some open questions from @khaosdoctor here. I'm happy with what you have right now, but if you make more changes, let me know and I'll re-review.

File: scaler/handlers.go
Comment (ajanth97) [reply scaler/handlers.go:67]: Yeah you are right. Actually I did try playing around with that method and then finally just reverted the changes in that method but I thought to leave the same variable name. I understand the value is not used but just for the sake of consistency.

**Comment by ajanth97** [comment]
> @ajanth97 I see that there are still some open questions from @khaosdoctor here. I'm happy with what you have right now, but if you make more changes, let me know and I'll re-review.

@arschles Awesome ! These are all the changes, let's wait for @khaosdoctor and get it merged then ?

**Comment by arschles** [comment]
Yup, sounds good. Looking forward to next week @ajanth97 !

File: scaler/handlers.go
Comment (khaosdoctor) [reply scaler/handlers.go:51]: I think that, in the case of not setting the `hostCount` to 0, we wouldn't have a big problem. Because there were to be only two return statements, and they're both different from each other, since the first one is an early return to avoid scaling the interceptor down. The final idea I could think of was:

```go
func (e *impl) IsActive(
	ctx context.Context,
	scaledObject *externalscaler.ScaledObjectRef,
) (*externalscaler.IsActiveResponse, error) {
	lggr := e.lggr.WithName("IsActive")
	host, ok := scaledObject.ScalerMetadata["host"]
	if !ok {
    err := fmt.Errorf("no 'host' field found in ScaledObject metadata")
		lggr.Error(err, "returning immediately from IsActive RPC call", "ScaledObject", *scaledObject)
		return nil, err
	}

	allCounts := e.pinger.counts()
	active, err := isActive(host, allCounts)
	if err != nil {
		err := fmt.Errorf("host '%s' not found in counts", host)
		lggr.Error(err, "Given host was not found in queue count map", "host", host, "allCounts", allCounts)
		return nil, err
	}

	return &externalscaler.IsActiveResponse{
		Result: active,
	}, nil
}

func isActive (host string, allCounts map[string]int) (bool, error) {
	if host == "interceptor" {
		return true
	}

	hostCount, ok := allCounts[host]
	if !ok {
		return false, fmt.Errorf("host '%s' not found in counts", host)
	}

	return hostCount > 0, nil
}
```

But this doesn't add much, so it's better to stick with this implementation.

File: scaler/handlers.go
Comment (khaosdoctor) [reply scaler/handlers.go:58]: Agree with that

**Comment by khaosdoctor** [comment]
Approving here :) Thanks @ajanth97!

**Review by khaosdoctor — APPROVED** [review]

**Comment by khaosdoctor** [comment]
> > @ajanth97 I see that there are still some open questions from @khaosdoctor here. I'm happy with what you have right now, but if you make more changes, let me know and I'll re-review.
> 
> @arschles Awesome ! These are all the changes, let's wait for @khaosdoctor and get it merged then ?

Sorry to keep you all waiting!

**Comment by arschles** [comment]
Thanks @ajanth97 for this! And @khaosdoctor thanks for the review :)

Merging now

**Comment by ajanth97** [comment]
Thank you @arschles and @khaosdoctor

**Pull request merged**
